### ISSUE TITLE
Set up the ML and Enable it to Feed Data to be used for front end

### ISSUE THREAD

**Issue by Meo-exe** [Meo-exe/push-push#13]
#Setting up the ML

## Dataframe building

- [x] Set up a function that will import the data from the Supabase
- [x] Create the database as a session so that it doesn't have to be stored locally, saving a lot of space
- [x] Create a function that will be able to add the data from the latest race as well as populate independent variables that will be used to make the predictions of the next race
- [x] Determine a best time to trigger the importation of new data and prediction. 

## Prediction

- [x] Train the model that will make predictions using the data from the Supabase. 
- [ ] Try different models and determine which will be the best for our use case. Important things to consider will be server capabilities since it is likely to be the biggest bottleneck. 
- [ ] Make the prediction and return an output that will be used by backend to display predictions to frontend. 

## Routing

- [ ] Creating the routing and systems that will be needed so that the ML model can interact with the rest of the backend and front end.

**Comment by Meo-exe** [comment]
##Dataframe Building
Most of the dataframe building is complete. Currently able to import the data using the Result Database. Since the Result Database had most of the useful independent variable it was the obvious choice. Some changes were made to the models to accommodate for the backwards population into the the Results from other Datasets (Drivers, Races, Teams). **SQLAlchemy** has a _**relationship**_ module integrated into it to link all of our models with one another. All of the dataframe building is Session based using our **_get_db_** function so as to conserve space. 
To be able to easily access which race is the next, a function goes through a loop of the results for the current year and compares the race_id ('{year}-{round}', ex: '2025-6') in races database, which holds the race_id of every race for that year, and the race_ids present in the results database. The first one that is not found in the results is determined to be the next race. 

- [ ] A current is that our API doesn't offer a way to query for drivers that will participate in the upcoming races to make a prediction. As of right now, there are two options: To use the drivers from the last race, or to manually change drivers whenever a roster change is announced. Looking into finding a better API could be a potential solution in the future. 

Currently, a new prediction is triggered on the Monday of each week.

**Comment by Meo-exe** [comment]
## Predictions
The model is currently able to output predictions. It is doing so by giving a probability of winning the race of each of the drivers and then ranking them based on their odds from 1-20. Some small adjustments were made to ensure that no two drivers would be predicted to end in the same positions. 
So far a very simple model has been used **RandomForestClassifier** as it was the simplest to implement at test for a first run. IT is likely that it will remain the model that will be used for this project as it has very low requirements. However, it would be nice to explore other options. Some suggested models are:

- [ ] XGBoost/LightGBM (Gradient Boosting/Classifier)
- [ ] MLP (probably overkill)
